### PR TITLE
Remove warnings from Ruby (MRI)

### PR DESCRIPTION
**Pull request by AlexWayfer**
Also enable them for RSpec.

### CODE REVIEW

**Comment by mdub** [comment]
Can you say more about why you think this is necessary?

**Comment by AlexWayfer** [comment]
> Can you say more about why you think this is necessary?

Just with `--warnings` flag for RSpec and without other changes tests looks like:

```
> bundle exec rspec
/home/alex/Projects/ruby/clamp/lib/clamp/subcommand/definition.rb:14: warning: method redefined; discarding old description
/home/alex/Projects/ruby/clamp/lib/clamp/subcommand/definition.rb:14: warning: method redefined; discarding old subcommand_class
........................................................................................................................../home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
./home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
../home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
/home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
........./home/alex/Projects/ruby/clamp/lib/clamp/attribute/definition.rb:57: warning: instance variable @required not initialized
.../home/alex/Projects/ruby/clamp/lib/clamp/attribute/definition.rb:57: warning: instance variable @required not initialized
...../home/alex/Projects/ruby/clamp/lib/clamp/attribute/definition.rb:57: warning: instance variable @required not initialized
./home/alex/Projects/ruby/clamp/lib/clamp/attribute/definition.rb:57: warning: instance variable @required not initialized
/home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
/home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
./home/alex/Projects/ruby/clamp/lib/clamp/attribute/definition.rb:57: warning: instance variable @required not initialized
/home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
/home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
....../home/alex/Projects/ruby/clamp/lib/clamp/subcommand/declaration.rb:62: warning: instance variable @default_subcommand not initialized
/home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
/home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
/home/alex/Projects/ruby/clamp/lib/clamp/option/declaration.rb:32: warning: instance variable @implicit_options_declared not initialized
/home/alex/Projects/ruby/clamp/lib/clamp/attribute/definition.rb:61: warning: instance variable @hidden not initialized
/home/alex/Projects/ruby/clamp/lib/clamp/attribute/definition.rb:61: warning: instance variable @hidden not initialized
/home/alex/Projects/ruby/clamp/lib/clamp/attribute/definition.rb:57: warning: instance variable @required not initialized
/home/alex/Projects/ruby/clamp/lib/clamp/attribute/definition.rb:57: warning: instance variable @required not initialized
..../home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
/home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
/home/alex/Projects/ruby/clamp/lib/clamp/option/declaration.rb:32: warning: instance variable @implicit_options_declared not initialized
/home/alex/Projects/ruby/clamp/lib/clamp/attribute/definition.rb:61: warning: instance variable @hidden not initialized
/home/alex/Projects/ruby/clamp/lib/clamp/attribute/definition.rb:57: warning: instance variable @required not initialized
./home/alex/Projects/ruby/clamp/lib/clamp/option/declaration.rb:32: warning: instance variable @implicit_options_declared not initialized
/home/alex/Projects/ruby/clamp/lib/clamp/attribute/definition.rb:57: warning: instance variable @required not initialized
/home/alex/Projects/ruby/clamp/lib/clamp/attribute/definition.rb:57: warning: instance variable @required not initialized
.........../home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
........./home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
......./home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
/home/alex/Projects/ruby/clamp/lib/clamp/messages.rb:14: warning: too many arguments for format string
...

Finished in 0.05434 seconds (files took 0.08845 seconds to load)
185 examples, 0 failures
```

In other libraries (gems), which are using Clamp, there are may be more warnings, it depends on number of tests cases.

Why these warnings are needed… you can search or ask Ruby team.

For example: https://www.ruby-forum.com/t/warning-instance-variable-foo-not-initialized/179396/2

It seems obvious to me that using unassigned instance variable, which returns `nil` as falsy value.

File: lib/clamp/subcommand/definition.rb [hunk 1]
```diff
@@ -3,16 +3,13 @@
...
-        @description = description
-        @subcommand_class = subcommand_class
+        names = Array(names)
+        super
       end
 
-      attr_reader :names, :description, :subcommand_class
-
       def is_called?(name)
         names.member?(name)
```
Comment (shiramax) [inline lib/clamp/subcommand/definition.rb:14]: names is no longer a global parameter, so it will be nil here.

Comment (AlexWayfer) [reply lib/clamp/subcommand/definition.rb:14]: Why do you think so? And what do you mean about "global parameter"? Instance variable?

**Comment by mdub** [comment]
Thanks @AlexWayfer.

**Pull request merged**